monitoring/exporter/exporter.go

```diff
@@ -0,0 +1,232 @@
+	// this function is called, it's guaranteed that at least one row data is also passed to
+	// OnError. Row data passed to OnError must not be modified. When OnError is not set, all
+	// errors happened on exporting are ignored.
+	OnError func(error, ...*RowData)
+	// MakeResource creates monitored resource from RowData. It is guaranteed that only RowData
+	// that passes GetProjectID will be given to this function. Though not recommended, error
+	// can be returned, and in that case the error is reported to callers via OnError and the
+	// row data will not be uploaded to stackdriver. When MakeResource is not set, global
+	// resource is used for all RowData objects.
+	MakeResource func(rd *RowData) (*monitoredrespb.MonitoredResource, error)
+
+	// options concerning labels.
+
+	// DefaultLabels store default value of some labels. Labels in DefaultLabels need not be
+	// specified in tags of view data. Default labels and tags of view may have overlapping
+	// label keys. In this case, values in tag are used. Default labels are used for labels
+	// those are constant throughout export operation, like version number of the calling
+	// program.
+	DefaultLabels map[string]string
+	// UnexportedLabels contains key of labels that will not be exported stackdriver. Typical
+	// uses of unexported labels will be either that marks project ID, or that's used only for
+	// constructing resource.
+	UnexportedLabels []string
+}
+
+// default values for options
+func defaultGetProjectID(rd *RowData) (string, error) {
+	return "", RowDataNotApplicableError
+}
+
+func defaultOnError(err error, rds ...*RowData) {}
+
+func defaultMakeResource(rd *RowData) (*monitoredrespb.MonitoredResource, error) {
+	return &monitoredrespb.MonitoredResource{Type: "global"}, nil
+}
+
+// NewStatsExporter creates a StatsExporter object. Once a call to NewStatsExporter is made, any
+// fields in opts must not be modified at all. ctx will also be used throughout entire exporter
+// operation when making RPC call.
+func NewStatsExporter(ctx context.Context, opts *Options) (*StatsExporter, error) {
+	client, err := newMetricClient(ctx, opts.ClientOptions...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create a metric client: %v", err)
+	}
+
+	e := &StatsExporter{
+		ctx:         ctx,
+		client:      client,
+		opts:        opts,
+		projDataMap: make(map[string]*projectData),
+	}
+
+	// We don't want to modify user-supplied options, so save default options directly in
+	// exporter.
+	if opts.GetProjectID != nil {
+		e.getProjectID = opts.GetProjectID
+	} else {
+		e.getProjectID = defaultGetProjectID
+	}
+	if opts.OnError != nil {
+		e.onError = opts.OnError
+	} else {
+		e.onError = defaultOnError
+	}
+	if opts.MakeResource != nil {
+		e.makeResource = opts.MakeResource
+	} else {
+		e.makeResource = defaultMakeResource
+	}
+
+	return e, nil
+}
+
+// We wrap monitoring.MetricClient and it's maker for testing.
+type metricClient interface {
```

Comment:
You shouldn't have to define an interface for this. Instead the client member in the exporter should just be a monitoring.MetricClient object, and in your test you should make a fake MetricClient by mocking the MetricServiceClient in the below definition of MetricClient.

// MetricClient is a client for interacting with Stackdriver Monitoring API.
type MetricClient struct {
	// The connection to the service.
	conn *grpc.ClientConn

	// The gRPC API client.
	metricClient monitoringpb.MetricServiceClient

	// The call options for this service.
	CallOptions *MetricCallOptions

	// The metadata to be sent with each request.
	metadata metadata.MD
}

Reply:
ok. I removed that interface but mocked monitoring.NewMetricClient and monitoring.MetricClient.CreateTimesSeries